presto-main/src/main/java/io/prestosql/operator/HashSemiJoinOperator.java

```diff
@@ -78,118 +89,161 @@ public OperatorFactory duplicate()
-        checkState(!finishing, "Operator is finishing");
-        checkState(channelSet != null, "Set has not been built yet");
-        checkState(outputPage == null, "Operator still has pending output");
-
-        // create the block builder for the new boolean column
-        // we know the exact size required for the block
-        BlockBuilder blockBuilder = BOOLEAN.createFixedSizeBlockBuilder(page.getPositionCount());
-
-        Page probeJoinPage = new Page(page.getBlock(probeJoinChannel));
-        Optional<Block> hashBlock = probeHashChannel.map(page::getBlock);
-
-        // update hashing strategy to use probe cursor
-        for (int position = 0; position < page.getPositionCount(); position++) {
-            if (probeJoinPage.getBlock(0).isNull(position)) {
-                if (channelSet.isEmpty()) {
-                    BOOLEAN.writeBoolean(blockBuilder, false);
-                }
-                else {
-                    blockBuilder.appendNull();
-                }
+    }
+
+    private class SemiJoinPages
+            implements WorkProcessor.Transformation<Page, Page>
+    {
+        private final int probeJoinChannel;
+        private final ListenableFuture<ChannelSet> channelSetFuture;
+        private final Optional<Integer> probeHashChannel;
+        private final LocalMemoryContext localMemoryContext;
+
+        @Nullable
+        private ChannelSet channelSet;
+
+        public SemiJoinPages(SetSupplier channelSetFuture, int probeJoinChannel, Optional<Integer> probeHashChannel, AggregatedMemoryContext aggregatedMemoryContext)
+        {
+            checkArgument(probeJoinChannel >= 0, "probeJoinChannel is negative");
+
+            this.channelSetFuture = requireNonNull(channelSetFuture, "hashProvider is null").getChannelSet();
+            this.probeJoinChannel = probeJoinChannel;
+            this.probeHashChannel = requireNonNull(probeHashChannel, "hashChannel is null");
+            this.localMemoryContext = requireNonNull(aggregatedMemoryContext, "aggregatedMemoryContext is null").newLocalMemoryContext(SemiJoinPages.class.getSimpleName());
+        }
+
+        @Override
+        public TransformationState<Page> process(Page inputPage)
+        {
+            if (inputPage == null) {
+                return finished();
             }
-            else {
-                boolean contains;
-                if (hashBlock.isPresent()) {
-                    long rawHash = BIGINT.getLong(hashBlock.get(), position);
-                    contains = channelSet.contains(position, probeJoinPage, rawHash);
-                }
-                else {
-                    contains = channelSet.contains(position, probeJoinPage);
+
+            if (channelSet == null) {
+                if (!channelSetFuture.isDone()) {
+                    // This will materialize page but it shouldn't matter for the first page
+                    localMemoryContext.setBytes(inputPage.getSizeInBytes());
```

Comment:
@dain this makes `SemiJoinOperator` keep first input page (to progress probe computations before future is done). I don't think it's relevant.

Reply:
For more context we want to progress upstream operators computations (e.g aggregation) by pulling initial page from probe side when build side future is not done. In current operator model this is also the case. However, in current model computations are performed up to input of next upstream operator. In `WorkProcessorOperator` model computations are performed up to input of `HashSemiJoinOperator`.